### DISCUSSION TITLE
Is it possible to use wagmi without a global / static chain configuration?

### DISCUSSION THREAD

**Discussion by naps62** [wevm/wagmi#2178] [Q&A]
I can't find any example of how I'd do this

I have a project where users are able to create/remove RPC endpoints for EVM chains on the UI itself, and then connect to those
so the known chains are not known as a static value, but would need to be configurable within a react hook, or something similar, based on external input

Something along these lines (non-working pseudocode):

```typescript
const Context = createContext(null);

function App() {
  const [provider, setProvider] = useState<JsonRpcProvider|null>(null);

  const { data } = useSWR(...) // some external data that tells me what the current RPC URL should be

  useEffect(() => {
    const { provider } = configureChains( ??? )
    setProvider(provider);
  }, [data])

  return <MyApp />
}
```

Sidenote: I was trying to setup an actual prototype, but Typescript doesn't let me because there seems to be some discrepancy on the [docs](https://wagmi.sh/core/providers/jsonRpc#usage). They state that `configureChains` returns a `{ provider: JsonRpcProvider }`, but in fact it returns a `({ chainId }: { chainId?: number | undefined; }) => (FallbackProvider ... )`

**Comment by jxom** [comment]
You could probably manually construct your own definition of a dynamic chain that follows the `Chain` type and has your dynamic RPC URLs.

Note that, you need to pass the `provider` into wagmi’s `createClient` before you can use wagmi.

**Reply by naps62** [reply]
the part about creating my own `Chain` I think I can figure out

but are you then saying that I should be able to do that and call `configureChains` / `createClient` from within a react hook?

**Comment by naps62** [comment]
I was able to get this working (I think. so far so good)

WIll leave here for future reference, but could also use some feedback since I'm not 100% sure this will hold up:

```typescript
import { type ProviderWithFallbackConfig } from "@wagmi/core";
import { providers } from "ethers";
import { useEffect, useState } from "react";
import {
  Chain,
  Client,
  WagmiConfig,
  configureChains,
  createClient,
} from "wagmi";
import { jsonRpcProvider } from "wagmi/providers/jsonRpc";

import { useInvoke } from "../hooks/tauri";
import { Network } from "../types";

interface Props {
  children: React.ReactNode;
}

type WagmiChains = { chains: Chain[] };
type WagmiProvider =
  | providers.FallbackProvider
  | ProviderWithFallbackConfig<providers.JsonRpcProvider>;

type WagmiClient = Client<
  WagmiProvider & WagmiChains,
  providers.WebSocketProvider & WagmiChains
>;

export function WagmiWrapper({ children }: Props) {
  const { data: network } = useInvoke<Network>("get_current_network");
  const [client, setClient] = useState<WagmiClient>();

  useEffect(() => {
    if (!network) return;
    setClient(buildClient(network));
  }, [network]);

  if (!client) return <></>;

  return <WagmiConfig client={client}>{children}</WagmiConfig>;
}

function buildClient(network: Network) {
  const { provider, webSocketProvider } = configureChains(
    [buildChain(network)],
    [
      jsonRpcProvider({
        rpc: () => ({ http: network?.http_url }),
      }),
    ]
  );

  return createClient({ provider, webSocketProvider });
}

function buildChain(network: Network): Chain {
  const rpcs = {
    http: [network.http_url],
    ws: network.ws_url ? [network.ws_url] : [],
  };

  return {
    id: network.chain_id,
    network: network.name,
    name: network.name,
    nativeCurrency: {
      name: "Ether",
      symbol: network.currency,
      decimals: network.decimals,
    },
    rpcUrls: {
      default: rpcs,
      public: rpcs,
    },
  };
}

```
